src/app/modules/form/admin-form/admin-form.routes.ts

```diff
@@ -0,0 +1,497 @@
+
+import { VALID_UPLOAD_FILE_TYPES } from '../../../../shared/constants'
+import { IForm, ResponseMode } from '../../../../types'
+import { withUserAuthentication } from '../../auth/auth.middlewares'
+import * as EncryptSubmissionController from '../../submission/encrypt-submission/encrypt-submission.controller'
+
+import * as AdminFormController from './admin-form.controller'
+import { DuplicateFormBody } from './admin-form.types'
+
+export const AdminFormsRouter = Router()
+
+const Joi = BaseJoi.extend(JoiDate) as typeof BaseJoi
+
+// Validators
+const createFormValidator = celebrate({
+  [Segments.BODY]: {
+    form: Joi.object<Omit<IForm, 'admin'>>()
+      .keys({
+        // Require valid responsesMode field.
+        responseMode: Joi.string()
+          .valid(...Object.values(ResponseMode))
+          .required(),
+        // Require title field.
+        title: Joi.string().min(4).max(200).required(),
+        // Require emails string (for backwards compatibility) or string
+        // array if form to be created in Email mode.
+        emails: Joi.alternatives()
+          .try(Joi.array().items(Joi.string()).min(1), Joi.string())
+          .when('responseMode', {
+            is: ResponseMode.Email,
+            then: Joi.required(),
+          }),
+        // Require publicKey field if form to be created in Storage mode.
+        publicKey: Joi.string()
+          .allow('')
+          .when('responseMode', {
+            is: ResponseMode.Encrypt,
+            then: Joi.string().required().disallow(''),
+          }),
+      })
+      .required()
+      // Allow other form schema keys to be passed for form creation.
+      .unknown(true),
+  },
+})
+
+const duplicateFormValidator = celebrate({
+  [Segments.BODY]: Joi.object<DuplicateFormBody>({
+    // Require valid responsesMode field.
+    responseMode: Joi.string()
+      .valid(...Object.values(ResponseMode))
+      .required(),
+    // Require title field.
+    title: Joi.string().min(4).max(200).required(),
+    // Require emails string (for backwards compatibility) or string array
+    // if form to be duplicated in Email mode.
+    emails: Joi.alternatives()
+      .try(Joi.array().items(Joi.string()).min(1), Joi.string())
+      .when('responseMode', {
+        is: ResponseMode.Email,
+        then: Joi.required(),
+      }),
+    // Require publicKey field if form to be duplicated in Storage mode.
+    publicKey: Joi.string()
+      .allow('')
+      .when('responseMode', {
+        is: ResponseMode.Encrypt,
+        then: Joi.string().required().disallow(''),
+      }),
+  }),
+})
+
+const fileUploadValidator = celebrate({
+  [Segments.BODY]: {
+    fileId: Joi.string().required(),
+    fileMd5Hash: Joi.string().base64().required(),
+    fileType: Joi.string()
+      .valid(...VALID_UPLOAD_FILE_TYPES)
+      .required(),
+  },
+})
+
+AdminFormsRouter.route('/adminform')
+  // All HTTP methods of route protected with authentication.
+  .all(withUserAuthentication)
```

Comment:
is it possible to just use `AdminFormsRouter.use(withUserAuthentication)` at the top?

Reply:
Not possible, due to our non-RESTful API hierarchy :(.

See d741b76;

> apparently the way express handles pathing is that they flatten all paths; and since AdminFormRouter (due to lack of RESTfulness) is on the root route '/', the withUserAuthentication middleware is now applied across the application, which is not ideal.

Reply:
ahh I see